File: samples/java/DeliveryDispatch/Server/CourierSession/src/main/java/systems/zlink/samples/deliverydispatch/server/couriersession/CourierSessionApplication.java
package systems.zlink.samples.deliverydispatch.server.couriersession;

import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.boot.context.properties.EnableConfigurationProperties;
import org.springframework.context.annotation.Bean;

import systems.zlink.contracts.core.RoutingId;
import systems.zlink.framework.configuration.ZLinkMeshNodeBuilder;
import systems.zlink.framework.configuration.ZLinkMessageFlowLogMode;
import systems.zlink.framework.locations.redis.ZLinkRedisLocationStore;
import systems.zlink.framework.monitoring.ZLinkRouteMeshRuntime;
import systems.zlink.framework.spring.EnableZLinkFramework;
import systems.zlink.framework.spring.ZLinkFrameworkConfigurer;
import systems.zlink.samples.deliverydispatch.server.configuration.DeliveryDispatchReadinessReporter;
import systems.zlink.samples.deliverydispatch.server.configuration.SampleApplication;
import systems.zlink.samples.deliverydispatch.server.configuration.SampleLocationStore;
import systems.zlink.samples.deliverydispatch.server.configuration.SampleNames;
import systems.zlink.samples.deliverydispatch.server.configuration.SampleTopology;
import systems.zlink.samples.deliverydispatch.server.couriersession.sessions.CourierSession;

@EnableZLinkFramework
@EnableConfigurationProperties(SampleTopology.class)
@SpringBootApplication(
        proxyBeanMethods = false,
        scanBasePackageClasses = CourierSessionApplication.class)
public final class CourierSessionApplication {
    private CourierSessionApplication() {}

    public static AutoCloseable run(String configPath) {
        return SampleApplication.start(CourierSessionApplication.class, configPath)::close;
    }

    @Bean
    ZLinkFrameworkConfigurer courierSessionFramework(SampleTopology topology) {
        return options -> {
            options.addHandlersFromPackageOf(CourierSessionApplication.class);
            options.configureDispatch().messageFlow(ZLinkMessageFlowLogMode.NORMAL);
            options.addClientServerChannel(SampleNames.CourierChannel).client();
            ZLinkMeshNodeBuilder node = options.addRouteMesh(SampleNames.CourierSpotDiscovery);
            node.listen(topology.courierSessionSpotEndpoint())
                    .setRoutingId(RoutingId.from(SampleNames.CourierSessionNode));
            node.objects().client();
            options.addStreamNode(SampleNames.CourierStreamNode)
                    .bind(topology.courierStreamEndpoint())
                    .enableActorDispatch()
                    .registerSession(CourierSession.class);
        };
    }

    @Bean(destroyMethod = "close")
    ZLinkRedisLocationStore locationStore(SampleTopology topology) {
        return SampleLocationStore.create(topology);
    }

    @Bean
    DeliveryDispatchReadinessReporter readinessReporter(ZLinkRouteMeshRuntime routeMeshRuntime) {
        return new DeliveryDispatchReadinessReporter(
                routeMeshRuntime, SampleNames.CourierSessionNode, SampleNames.CourierSpotDiscovery);
    }
}

File: samples/kotlin/DeliveryDispatch/Server/CourierSession/src/main/kotlin/systems/zlink/samples/kotlin/deliverydispatch/server/couriersession/CourierSessionApplication.kt
package systems.zlink.samples.kotlin.deliverydispatch.server.couriersession

import kotlinx.coroutines.Dispatchers
import org.springframework.boot.WebApplicationType
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.boot.builder.SpringApplicationBuilder
import org.springframework.context.annotation.Bean
import systems.zlink.framework.configuration.ZLinkMessageFlowLogMode
import systems.zlink.framework.kotlin.useCoroutineHandlers
import systems.zlink.framework.locations.redis.ZLinkRedisLocationStore
import systems.zlink.framework.monitoring.ZLinkRouteMeshRuntime
import systems.zlink.framework.spring.EnableZLinkFramework
import systems.zlink.framework.spring.ZLinkFrameworkConfigurer
import systems.zlink.samples.kotlin.deliverydispatch.server.configuration.DeliveryDispatchReadinessReporter
import systems.zlink.samples.kotlin.deliverydispatch.server.configuration.SampleLocationStore
import systems.zlink.samples.kotlin.deliverydispatch.server.configuration.SampleNames
import systems.zlink.samples.kotlin.deliverydispatch.server.configuration.SampleTopology
import systems.zlink.samples.kotlin.deliverydispatch.server.couriersession.sessions.CourierSession

@EnableZLinkFramework
@SpringBootApplication(
    proxyBeanMethods = false,
    scanBasePackageClasses = [CourierSessionApplication::class],
)
class CourierSessionApplication {
    @Bean
    fun courierSessionFramework(): ZLinkFrameworkConfigurer = ZLinkFrameworkConfigurer { options ->
        options.useCoroutineHandlers(Dispatchers.Default)
        options.addHandlersFromPackageOf(CourierSessionApplication::class.java)
        options.configureDispatch().messageFlow(ZLinkMessageFlowLogMode.NORMAL)

        options.addClientServerChannel(SampleNames.CourierChannel).client()
        val node = options.addRouteMesh(SampleNames.CourierSpotMesh)
        node
            .listen(SampleTopology.CourierSessionSpotEndpoint)
            .setRoutingIdPrefix("delivery-session")
        node.objects().client()
        options
            .addStreamNode(SampleNames.CourierStreamNode)
            .bind(SampleTopology.CourierStreamEndpoint)
            .enableActorDispatch()
            .registerSession(CourierSession::class.java)
    }

    @Bean fun locationStore(): ZLinkRedisLocationStore = SampleLocationStore.create()

    @Bean(destroyMethod = "close")
    fun readinessReporter(meshes: ZLinkRouteMeshRuntime): DeliveryDispatchReadinessReporter =
        DeliveryDispatchReadinessReporter.route(
            SampleNames.CourierSessionNode,
            SampleNames.CourierSpotMesh,
            meshes,
        )

    companion object {
        fun run(args: Array<String> = emptyArray()): AutoCloseable {
            val builder =
                SpringApplicationBuilder(CourierSessionApplication::class.java)
                    .web(WebApplicationType.NONE)
            builder.application().setKeepAlive(true)
            val context = builder.run(*args)
            return AutoCloseable { context.close() }
        }
    }
}
